{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# nbdocker: quick demo\n",
    " \n",
    "### By: Jiaming Hu, Ling-Hong Hung and Ka Yee Yeung. \n",
    "\n",
    "Institute of Technology, University of Washington Tacoma, WA 98402, USA\n",
    "\n",
    "This is a quick demo of nbdocker, showing the feasibility of embedding a Docker container inside a Jupyter notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "In this quick demo, we will download fastq files from NCBI Short Read Archive (SRA) and then perform quality assessment using a R/Bioconductor package called ShortRead.  \n",
    "\n",
    "**This demo shows the following:**\n",
    "1. Docker containers can be embedded inside Jupyter notebooks using nbdocker. In addition, the history of the embedded containers are included in the notebook.  Sharing the notebook (.ipynb) files will include the history of Docker commands.\n",
    "2. The users can  can run and monitor the progress of the embedded Docker containers by a single click.\n",
    "3. We can mix and match computing environments and programming langugages. In particular, the download step involves shell commands and Python wrapper, while the quality control step uses R.\n",
    "4. Allowing multiple computing environment facilitates collaboration and reproducibility.\n",
    "\n",
    "**To run this demo nbdocker notebook:**\n",
    "1. Pull the nbdocker image from DockerHub: \n",
    ">docker pull biodepot/nbdocker\n",
    "2. Run the nbdocker image. The following command works on the mac.\n",
    ">docker run -it -p 8888:8888 -v /var/run/docker.sock:/var/run/docker.sock -v /Users/username/nbdocker:/home/jovyan/work --privileged --group-add root biodepot/nbdocker\n",
    "\n",
    "The above command maps the local directory */Users/username/nbdocker* on the host to */home/jovyan/work* inside the container running nbdocker.  Replace */Users/username/nbdocker* with the local directory that you'd like to map to your notebook.  Note that */home/jovyan/work* is equivalent to */.nbdocker* within the nbdocker notebook.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "DockerContainers": {
     "0": {
      "id": "82ae4bbddadf",
      "status": "running..."
     }
    }
   },
   "source": [
    "## 1. Download the fastq input files from SRA \n",
    "#### Note: running this container takes 5-20 minutes depending on your bandwidth.\n",
    "\n",
    "We use RNA-seq data generated by Trapnell et al. (Nat Biotechnol. 2013 Jan; 31(1): 10.1038/nbt.2450) in which they studied the differential gene expression of lung fibroblasts in response to loss of the developmental transcription factor HOXA1.  In this case study, we will compare changes in gene expression  in response to HOXA1 knockdown. \n",
    "\n",
    "The fastq files are publicly available from the Short Read Archive (SRA) with the following link:\n",
    "https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR493366\n",
    "\n",
    "To download the files we use the parallel-fastq-dump container. This is a python wrapper around the fastq-dump utility that allows it to download separate chunks of the files separately.\n",
    "\n",
    "We assume you have created the \"kallisto\" directory on your local host.  Since we have mapped the local directory */Users/username/nbdocker* on the host to */home/jovyan/work* inside the container running nbdocker, we assume the path */Users/username/nbdocker/kallisto* on your local host exists.  The fastq files will be downloaded to '/home/jovyan/work/kallisto' within the container running nbdocker.\n",
    "\n",
    "**Click on the whale button** below to run this container\n",
    "\n",
    "{nbdocker#0}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Perform quality assessment using a R/Bioconductor package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.6 (BiocInstaller 1.28.0), ?biocLite for help\n",
      "A new version of Bioconductor is available after installing the most recent\n",
      "  version of R; see http://bioconductor.org/install\n",
      "BioC_mirror: https://bioconductor.org\n",
      "Using Bioconductor 3.6 (BiocInstaller 1.28.0), R 3.4.3 (2017-11-30).\n",
      "Installing package(s) ‘ShortRead’\n",
      "Updating HTML index of packages in '.Library'\n",
      "Making 'packages.html' ... done\n",
      "Old packages: 'cluster', 'curl', 'devtools', 'evaluate', 'foreign', 'git2r',\n",
      "  'glue', 'inline', 'IRdisplay', 'MASS', 'Matrix', 'mgcv', 'mvtnorm', 'nlme',\n",
      "  'pbdZMQ', 'Rcpp', 'RcppArmadillo', 'RCurl', 'repr', 'robustbase', 'rrcov',\n",
      "  'stringi', 'stringr', 'survival'\n"
     ]
    }
   ],
   "source": [
    "# Install the \"ShortRead\" Bioconductor package\n",
    "source('http://bioconductor.org/biocLite.R')\n",
    "biocLite(\"ShortRead\")\n",
    "library (ShortRead)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'output'</li>\n",
       "\t<li>'SRR493366_1.fastq.gz'</li>\n",
       "\t<li>'SRR493366_2.fastq.gz'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'output'\n",
       "\\item 'SRR493366\\_1.fastq.gz'\n",
       "\\item 'SRR493366\\_2.fastq.gz'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'output'\n",
       "2. 'SRR493366_1.fastq.gz'\n",
       "3. 'SRR493366_2.fastq.gz'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"output\"               \"SRR493366_1.fastq.gz\" \"SRR493366_2.fastq.gz\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# list the fastq files already downloaded\n",
    "list.files('/home/jovyan/work/kallisto')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: FastqQA(10)\n",
       "QA elements (access with qa[[\"elt\"]]):\n",
       "  readCounts: data.frame(2 3)\n",
       "  baseCalls: data.frame(2 5)\n",
       "  readQualityScore: data.frame(1024 4)\n",
       "  baseQuality: data.frame(190 3)\n",
       "  alignQuality: data.frame(2 3)\n",
       "  frequentSequences: data.frame(100 4)\n",
       "  sequenceDistribution: data.frame(88 4)\n",
       "  perCycle: list(2)\n",
       "    baseCall: data.frame(1009 4)\n",
       "    quality: data.frame(7324 5)\n",
       "  perTile: list(2)\n",
       "    readCounts: data.frame(0 4)\n",
       "    medianReadQualityScore: data.frame(0 4)\n",
       "  adapterContamination: data.frame(2 1)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# perform quality assessment on the downloaded fastq files\n",
    "qaSummary <- qa('/home/jovyan/work/kallisto')\n",
    "qaSummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dest, recursive = TRUE):\n",
      "“'/home/jovyan/work/kallisto' already exists”"
     ]
    },
    {
     "data": {
      "text/html": [
       "'/home/jovyan/work/kallisto/index.html'"
      ],
      "text/latex": [
       "'/home/jovyan/work/kallisto/index.html'"
      ],
      "text/markdown": [
       "'/home/jovyan/work/kallisto/index.html'"
      ],
      "text/plain": [
       "[1] \"/home/jovyan/work/kallisto/index.html\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'/home/jovyan/work'"
      ],
      "text/latex": [
       "'/home/jovyan/work'"
      ],
      "text/markdown": [
       "'/home/jovyan/work'"
      ],
      "text/plain": [
       "[1] \"/home/jovyan/work\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# generate the assessment report\n",
    "report(qaSummary, dest=\"/home/jovyan/work/kallisto\")\n",
    "getwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The assessment report is generated as an index.html file in the demo directory. You can click on the following link to view it in the browser.\n",
    "\n",
    "[View FastQC Report](./kallisto/index.html)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "cmd_history": [
   {
    "Id": 0,
    "command": "bash -c 'parallel-fastq-dump --sra-id SRR493366 --threads 16 --outdir /.nbdocker/kallisto --split-files --gzip'",
    "container": "/.nbdocker",
    "external": "",
    "host": "/home/lhhung/nbdocker",
    "image": "biodepot/parallel-fastq-dump:latest",
    "internal": "",
    "notebookname": "nbdocker_demo"
   },
   {
    "Id": 1,
    "command": "bash -c 'parallel-fastq-dump --sra-id SRR493366 --threads 4 --outdir /.nbdocker/kallisto --split-files --gzip'",
    "container": "/.nbdocker",
    "external": "",
    "host": "/home/lhhung/nbdocker",
    "image": "biodepot/parallel-fastq-dump:latest",
    "internal": "",
    "notebookname": "nbdocker_demo"
   }
  ],
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
